### PR TITLE
TRAIT_TABLE_LEAP Checks Added To under_siege  Judo

### LINKED ISSUE #23766: `TRAIT_TABLE_LEAP` not retained upon cloning, but Corporate Judo is.
### Exploit Reports

- [X] I confirm this issue is not an exploit. (Required)

### BYOND Version

515.1621

### Issue Description

Upon cloning, bartender retains corporate judo as 	`var/datum/martial_art/judo/under_siege/bouncer_delight = new`, but not `TRAIT_TABLE_LEAP`

### What did you expect to happen?

Retain both corporate judo and table leap upon cloning..

### What happened instead?

Only retains table leap upon cloning.

### Why is this bad/What are the consequences?

Bartender unable to retain core gameplay strategy.

### Steps to reproduce the issue.

Die
Be Cloned
Attempt to leap table

### When did the problem start happening?

Round ID 38759

### Extra information

_No response_

### Relevant log output/runtime error

_No response_

### PR DESCRIPTION
**Pull request by Varelion**
## What Does This PR Do

This PR closes #23766. 
It checks for the table_leap trait in corporate judo.

## Why It's Good For The Game
Consistency in logic. Corporate judo is inherited through cloning; it stands to reason that table-jumping would as well. Due to where these things are stored, and how cloning works, previously you would only inherit corporate judo. Now, however, you inherit both; this is done by intergrading the ability to jump tables fast into Bartender's corporate judo.

## Testing
 Hosted a test server, got some members to test cloning with bartender, and their judo abilities before and after. 

## Changelog
:cl:
tweak: Add 'TRAIT_TABLE_LEAP' checks to 'under_siege' judo.
fix: closes #23766 
/:cl:

### CODE REVIEW

File: code/modules/martial_arts/judo.dm [hunk 1]
```diff
@@ -77,14 +77,20 @@
 
 /datum/martial_art/judo/under_siege/teach(mob/living/carbon/human/H, make_temporary)
 	RegisterSignal(H, COMSIG_AREA_ENTERED, PROC_REF(bar_check))
+	if(HAS_TRAIT(H, TRAIT_TABLE_LEAP) == FALSE)
```
Comment (DGamerL) [inline code/modules/martial_arts/judo.dm:80]:
```suggestion
	if(!HAS_TRAIT(H, TRAIT_TABLE_LEAP))
```

File: code/modules/martial_arts/judo.dm [hunk 2]
```diff
@@ -77,14 +77,20 @@
...
+		ADD_TRAIT(H, TRAIT_TABLE_LEAP,ROUNDSTART_TRAIT)
 	return ..()
 
 /datum/martial_art/judo/under_siege/remove(mob/living/carbon/human/H)
 	UnregisterSignal(H, COMSIG_AREA_ENTERED)
+	if(HAS_TRAIT(H, TRAIT_TABLE_LEAP) == TRUE)
```
Comment (DGamerL) [inline code/modules/martial_arts/judo.dm:86]:
```suggestion
	if(HAS_TRAIT(H, TRAIT_TABLE_LEAP))
```

File: code/modules/martial_arts/judo.dm [hunk 3]
```diff
@@ -77,14 +77,20 @@
...
+		REMOVE_TRAIT(H, TRAIT_TABLE_LEAP,ROUNDSTART_TRAIT)
 	return ..()
 
 /datum/martial_art/judo/under_siege/proc/bar_check(mob/living/carbon/human/H, area/entered_area)
 	SIGNAL_HANDLER
+	if(HAS_TRAIT(H, TRAIT_TABLE_LEAP) == FALSE)
```
Comment (DGamerL) [inline code/modules/martial_arts/judo.dm:92]:
```suggestion
	if(!HAS_TRAIT(H, TRAIT_TABLE_LEAP))
```

File: code/modules/martial_arts/judo.dm [hunk 4]
```diff
@@ -97,6 +103,8 @@
 
 /datum/martial_art/judo/under_siege/can_use(mob/living/carbon/human/H)
 	var/area/A = get_area(H)
+	if(HAS_TRAIT(H, TRAIT_TABLE_LEAP) == FALSE)
```
Comment (DGamerL) [inline code/modules/martial_arts/judo.dm:106]:
```suggestion
	if(!HAS_TRAIT(H, TRAIT_TABLE_LEAP))
```

**Review by DGamerL — COMMENTED** [review]
While this does fix the issue for the bartender, this doesn't fix it for the chef.
You could consider making it a mind trait instead? That way it will transfer upon cloning

**Comment by Henri215** [comment]
Change your changelog to something like "fix: bartenders no longer loses their jumping over table skills when cloned". Remember that it should be written considering the average player is going to read it and must understand it.

**Review by Contrabang — CHANGES_REQUESTED** [review]
Adding the trait onto the martial art is really hacky, instead it should be done through signals or having the trait be on the mind.

Using something like this on a human, will transfer the trait
`RegisterSignal(src, COMSIG_BODY_TRANSFER_TO, PROC_REF(make_a_new_proc))`

```
/mob/living/carbon/human/proc/whatever_proc_name()
    ADD_TRAIT(H, TRAIT_TABLE_LEAP, ROUNDSTART_TRAIT)
```

**Comment by DGamerL** [comment]
> Adding the trait onto the martial art is really hacky, instead it should be done through signals or having the trait be on the mind.
> 
> Using something like this on a human, will transfer the trait `RegisterSignal(src, COMSIG_BODY_TRANSFER_TO, PROC_REF(make_a_new_proc))`
> 
> ```
> /mob/living/carbon/human/proc/whatever_proc_name()
>     ADD_TRAIT(H, TRAIT_TABLE_LEAP, ROUNDSTART_TRAIT)
> ```

I'd suggest just doing it on the mind yeah, much easier

**Comment by Varelion** [comment]
> Adding the trait onto the martial art is really hacky, instead it should be done through signals or having the trait be on the mind.
> 
> Using something like this on a human, will transfer the trait `RegisterSignal(src, COMSIG_BODY_TRANSFER_TO, PROC_REF(make_a_new_proc))`
> 
> ```
> /mob/living/carbon/human/proc/whatever_proc_name()
>     ADD_TRAIT(H, TRAIT_TABLE_LEAP, ROUNDSTART_TRAIT)
> ```

I'm looking into the mind implementation, but I'd also be introducing the functionality to add and remove these admin side through the procs present in CQC and Judo's respective siege sides. Why would I go about it through a signal, and not an if statement?

File: code/__HELPERS/trait_helpers.dm [hunk 5]
```diff
@@ -226,6 +226,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 
 //***** MIND TRAITS *****/
 #define TRAIT_HOLY "is_holy" // The mob is holy in regards to religion
+#define TRAIT_TABLE_LEAP "table_leap"
+
```
Comment (DGamerL) [inline code/__HELPERS/trait_helpers.dm:230]:
```suggestion
```

**Review by DGamerL — APPROVED** [review]
Could've sworn we had a helper to add to immediately add to a mob's mind, but LGTM

**Review by Sirryan2002 — APPROVED** [review]

**Review by Contrabang — APPROVED** [review]

**Review by S34NW — APPROVED** [review]

**Pull request merged**